### PR TITLE
launch: add launch specific idle throttle param

### PR DESCRIPTION
**Pull request by tstastny**
Depending on the platform, fixed-wing hand launches may require non-zero idle throttle while "in hand" and getting a running start (e.g. to push more prop-wash over the control surfaces while running slower than the stall speed). This conflicts with the idle throttle setting engaged after land detection, which, in this case, would then keep idling the motor at that same non-zero value, when for e.g. belly landings this is not desirable (should be set to zero).

This pr adds a pre-takeoff idle throttle for catapult (or hand launch) methods, which differs from the idle throttle set after land detection. It is only active until motors are enabled by the launch detection -- which then as before initiates climb out.

Current implementation seems to work testing outside on the bench. But maybe there are some recommendations on how better to structure the changes in the fw pos ctrl cpp, @dagar ?

### CODE REVIEW

**Comment by Antiheavy** [comment]
@tstastny I'm just curious, do you plan to use this for a hand launched airplane?  I definitely agree that throttle during hand launch motion can be critical to hand launch success (our plane is this way), but we use LaunchDetection to start the motor by shaking the vehicle to trip the accelerator, then throw.  Do you have a button or something on the vehicle to arm and start the motor?

**Comment by Antiheavy** [comment]
neat.  Hand launch controller and vehicle design is harder than most people think.  best of luck.

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.hpp [hunk 1]
```diff
@@ -418,7 +418,7 @@ class FixedwingPositionControl final : public ModuleBase<FixedwingPositionContro
 	bool		control_position(const Vector2f &curr_pos, const Vector2f &ground_speed, const position_setpoint_s &pos_sp_prev,
 					 const position_setpoint_s &pos_sp_curr);
 	void		control_takeoff(const Vector2f &curr_pos, const Vector2f &ground_speed, const position_setpoint_s &pos_sp_prev,
-					const position_setpoint_s &pos_sp_curr);
+					const position_setpoint_s &pos_sp_curr, float &pre_takeoff_throttle);
```
Comment (dagar) [inline src/modules/fw_pos_control_l1/FixedwingPositionControl.hpp]: I might not be following this correctly, but why do we need to pass this through to the takeoff routine?

Comment (tstastny) [reply src/modules/fw_pos_control_l1/FixedwingPositionControl.hpp]: Good point. I'm not sure what my original thought process was there. -- will change it and just call the throttle idle get from launch detection whenever it's needed.

**Comment by philipoe** [comment]
@dagar @tstastny Update? Is this good to merge given that all review comments were addressed?

**Comment by philipoe** [comment]
@dagar Can we merge, given that checks have passed and this was foreseen for Release 1.8.0 originally?

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp [hunk 2]
```diff
@@ -1060,7 +1060,7 @@ FixedwingPositionControl::control_position(const Vector2f &curr_pos, const Vecto
 		/* making sure again that the correct thrust is used,
 		 * without depending on library calls for safety reasons.
 		   the pre-takeoff throttle and the idle throttle normally map to the same parameter. */
-		_att_sp.thrust = _parameters.throttle_idle;
+		_att_sp.thrust = _launchDetector.getThrottleIdle(_parameters.throttle_idle);
```
Comment (dagar) [inline src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp]: Typically the pattern here is to pass the value you'd use in the non-launch detection case, and launch detector would return one or the other depending on if it's enabled.

Comment (tstastny) [reply src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp]: I believe this is what is happening. Or do I miss something? (at least this is what I intended)

[LaunchDetector.cpp#L128](https://github.com/tstastny/Firmware/blob/6cb7bc880f9c66b660d68bef96b730019427dbac/src/modules/fw_pos_control_l1/launchdetection/LaunchDetector.cpp#L128)

Comment (dagar) [reply src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp]: My mistake, that looks correct.

**Comment by tstastny** [comment]
> the merits for and against forcing LandDector into the Landed state while Launch Detection is running

@Antiheavy right -- this was actually something we just found tonight testing in HIL -- will check if this fix works tomorrow morning, but it seems like quite a few corner cases with the landing state. hopefully this commit works, but of course there are probably other (possibly better) ways to handle this more generally.

**Comment by tstastny** [comment]
missed (didnt add locally) the latter part of the last commit on the last push -- now works as advertised. see https://review.px4.io/plot_app?log=02bd7a15-762f-4118-997b-1b468cce2391

@philipoe @dagar

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp [hunk 3]
```diff
@@ -1896,10 +1899,12 @@ FixedwingPositionControl::tecs_update_pitch_throttle(float alt_sp, float airspee
 	}
 
 	/* tell TECS to update its state, but let it know when it cannot actually control the plane */
-	bool in_air_alt_control = (!_vehicle_land_detected.landed &&
-				   (_control_mode.flag_control_auto_enabled ||
-				    _control_mode.flag_control_velocity_enabled ||
-				    _control_mode.flag_control_altitude_enabled));
+	bool in_air_alt_control = ((!_vehicle_land_detected.landed ||
+				    (_launch_detection_state == LAUNCHDETECTION_RES_DETECTED_ENABLEMOTORS
+				     && mode == tecs_status_s::TECS_MODE_TAKEOFF))
```
Comment (dagar) [inline src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp]: This is a mouthful, should we have a helper function? Same logic could be used for run_tecs above?

**Comment by dagar** [comment]
Instead of completely disregarding land detect in takeoff after launch detect what if we gave it a small window of a few seconds? At that point if there's not a successful transition to flying the land detector would kill the throttle again and the vehicle could even auto disarm. I'm thinking of various failed launch scenarios.

**Comment by tstastny** [comment]
> Instead of completely disregarding land detect in takeoff after launch detect what if we gave it a small window of a few seconds? At that point if there's not a successful transition to flying the land detector would kill the throttle again and the vehicle could even auto disarm. I'm thinking of various failed launch scenarios.

Sure, it is a valid concern. We actually used to have an additional launch detection initialization command on QGC (a button) on our custom firmware that did something similar. Basically you could switch to auto, but still needed to hit the button to start launch detection (simultaneously spooling up the motors and then ignoring land detection). This could then be coupled with a timeout as you say.

Is there an existing mavlink command that is best used for this purpose? maybe this one: https://mavlink.io/en/messages/common.html#MAV_CMD_NAV_TAKEOFF ? I guess this would require some QGC work as well. Otherwise one would need to flip out of auto and back in to reinitialize the launch detection after a land was detected.

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp (hunk 3 above)
Comment (tstastny) [reply src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp]: e.g. just an extra bool calculated directly above?
```
bool is_launching = _launch_detection_state == LAUNCHDETECTION_RES_DETECTED_ENABLEMOTORS && mode == tecs_status_s::TECS_MODE_TAKEOFF;
bool in_air_alt_control = !_vehicle_land_detected.landed || is_launching;
```

or do you mean an encapsulated function call?
```
bool in_air_alt_control = !_vehicle_land_detected.landed || is_launching();
```

**Comment by philipoe** [comment]
@dagar @tstastny Update? How shall we proceed here?

**Comment by Antiheavy** [comment]
A short timer for disabling land detection seems like it might work. I can’t think of a better solution. When would the timer start? Immediately upon LaunchDetect?

We’ve also seen bad effects of LandDetector switching in and out of Landed state during ground handling and leading up to LaunchDetect. I worry about future problems there.

**Comment by priseborough** [comment]
The linked issue 
https://github.com/PX4/Firmware/issues/10076 indicates that we either need to start running the TECS internal updates when launch is detected rather than wait for onGround to go false, or set pitch angle demand to the min climbout value until onGround goes true.

**Comment by fredowski** [comment]
Just to support the idea of the launch idle throttle: We use the idle throttle for a hand launch of the X-UAV Mini Talon. Without the idle throttle the switch to 100% throttle via the accelerometer was too late and the launch did not succeed. We did not even think what happens when we land cause we were so happy that the launch worked... We activated the idle throttle (approx. 50%)  by switching to mission mode. But of course the throttle should be 0% after landing...

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp (hunk 3 above)
Comment (julianoes) [reply src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp]: The first one is good, I'd say.

**Comment by LorenzMeier** [comment]
Thanks for your earlier contribution. Unfortunately as the project has overall evolved quite a bit, this change doesn't apply any more. Closing stale pull requests like this one is part of us working aggressively to bring down the PR review time so that we will be able to merge or reject PRs in the future in a much more timely fashion.

**Pull request closed without merging**